File: packages/fallow-agent/src/fallow_agent/modelcache/download.py
"""Resumable, incrementally-hashed blob streaming.

This module owns exactly one HTTP attempt. Retry/backoff and verification live
in :mod:`fallow_agent.modelcache.store`; here we only:

- rehash any pre-existing ``.part`` prefix once (it may predate this process),
- send a ``Range`` request to resume,
- honour 206 (append) vs 200 (server ignored Range -> restart from zero),
- stream to disk in fixed chunks, hashing as bytes arrive.
"""

import hashlib
from collections.abc import Mapping
from pathlib import Path
from typing import NamedTuple, Protocol

import httpx

from fallow_agent.modelcache.config import HTTP_OK, HTTP_PARTIAL_CONTENT


class DownloadResult(NamedTuple):
    """Outcome of one completed stream: the running hash and total byte count."""

    sha256: str
    size: int


class RetryableStatusError(Exception):
    """A blob request returned a status we neither append to nor restart from."""

    def __init__(self, status_code: int) -> None:
        super().__init__(f"unexpected blob status {status_code}")
        self.status_code = status_code


class _Hasher(Protocol):
    def update(self, data: bytes, /) -> None: ...

    def hexdigest(self) -> str: ...


def _seed_from_existing(hasher: _Hasher, part: Path, block_size: int) -> int:
    """Feed an existing ``.part`` prefix through ``hasher``; return its size.

    The partial file may have been written by an earlier process, so we cannot
    assume any in-memory hash state carried over — we rehash the prefix once.
    """
    if not part.exists():
        return 0
    size = 0
    with part.open("rb") as fh:
        while True:
            block = fh.read(block_size)
            if not block:
                break
            hasher.update(block)
            size += len(block)
    return size


def _resolve_disposition(status_code: int, existing: int) -> tuple[bool, int]:
    """Map (status, resume offset) to (append?, bytes already counted).

    206 -> append to the prefix. 200 -> restart from zero even if we asked for a
    range (the coordinator ignored it). Anything else is retryable.
    """
    if status_code == HTTP_PARTIAL_CONTENT:
        return True, existing
    if status_code == HTTP_OK:
        return False, 0
    raise RetryableStatusError(status_code)


async def stream_to_part(
    *,
    client: httpx.AsyncClient,
    url: str,
    headers: Mapping[str, str],
    part: Path,
    chunk_size: int,
) -> DownloadResult:
    """Perform one resume-aware download attempt into ``part``."""
    hasher = hashlib.sha256()
    existing = _seed_from_existing(hasher, part, chunk_size)
    req_headers = dict(headers)
    if existing:
        req_headers["Range"] = f"bytes={existing}-"

    async with client.stream("GET", url, headers=req_headers) as response:
        append, total = _resolve_disposition(response.status_code, existing)
        if not append:
            hasher = hashlib.sha256()  # restart / full: discard seeded prefix
        with part.open("ab") if append else part.open("wb") as fh:
            async for chunk in response.aiter_bytes(chunk_size):
                fh.write(chunk)
                hasher.update(chunk)
                total += len(chunk)

    return DownloadResult(sha256=hasher.hexdigest(), size=total)

File: go-agent/modelcache/download.go
package modelcache

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"hash"
	"io"
	"net/http"
	"os"
)

// downloadResult is the outcome of one completed stream: the running hash and
// total byte count.
type downloadResult struct {
	sha256 string
	size   int64
}

// retryableStatusError signals a blob request returned a status the download
// path neither appends to (206) nor restarts from (200).
type retryableStatusError struct {
	statusCode int
}

func (e *retryableStatusError) Error() string {
	return fmt.Sprintf("unexpected blob status %d", e.statusCode)
}

// seedFromExisting feeds an existing ".part" prefix through hasher and returns
// its size. The partial file may have been written by an earlier process, so we
// cannot assume any in-memory hash state carried over — we rehash the prefix
// once.
func seedFromExisting(hasher hash.Hash, part string, blockSize int) (int64, error) {
	fh, err := os.Open(part)
	if err != nil {
		if os.IsNotExist(err) {
			return 0, nil
		}
		return 0, err
	}
	defer fh.Close()

	buf := make([]byte, blockSize)
	var size int64
	for {
		n, readErr := fh.Read(buf)
		if n > 0 {
			hasher.Write(buf[:n])
			size += int64(n)
		}
		if readErr == io.EOF {
			break
		}
		if readErr != nil {
			return 0, readErr
		}
	}
	return size, nil
}

// resolveDisposition maps (status, resume offset) to (append?, bytes already
// counted). 206 appends to the prefix; 200 restarts from zero even if a range
// was requested (the coordinator ignored it); anything else is retryable.
func resolveDisposition(statusCode int, existing int64) (bool, int64, error) {
	switch statusCode {
	case httpPartialContent:
		return true, existing, nil
	case httpOK:
		return false, 0, nil
	default:
		return false, 0, &retryableStatusError{statusCode: statusCode}
	}
}

// streamToPart performs one resume-aware download attempt into part. It owns
// exactly one HTTP attempt; retry, backoff, and verification live in Store.
func streamToPart(
	ctx context.Context,
	client *http.Client,
	url string,
	headers map[string]string,
	part string,
	chunkSize int,
) (downloadResult, error) {
	hasher := sha256.New()
	existing, err := seedFromExisting(hasher, part, chunkSize)
	if err != nil {
		return downloadResult{}, err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return downloadResult{}, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	if existing > 0 {
		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", existing))
	}

	resp, err := client.Do(req)
	if err != nil {
		return downloadResult{}, err
	}
	defer resp.Body.Close()

	appendMode, total, err := resolveDisposition(resp.StatusCode, existing)
	if err != nil {
		return downloadResult{}, err
	}
	if !appendMode {
		hasher = sha256.New() // restart / full: discard the seeded prefix
	}

	flags := os.O_CREATE | os.O_WRONLY
	if appendMode {
		flags |= os.O_APPEND
	} else {
		flags |= os.O_TRUNC
	}
	fh, err := os.OpenFile(part, flags, 0o644)
	if err != nil {
		return downloadResult{}, err
	}

	written, copyErr := streamChunks(fh, resp.Body, hasher, chunkSize)
	total += written
	if closeErr := fh.Close(); closeErr != nil && copyErr == nil {
		copyErr = closeErr
	}
	if copyErr != nil {
		return downloadResult{}, copyErr
	}

	return downloadResult{sha256: hex.EncodeToString(hasher.Sum(nil)), size: total}, nil
}

// streamChunks copies body into fh in fixed-size chunks, updating hasher as
// bytes arrive, and returns the number of bytes written.
func streamChunks(fh io.Writer, body io.Reader, hasher hash.Hash, chunkSize int) (int64, error) {
	buf := make([]byte, chunkSize)
	var total int64
	for {
		n, readErr := body.Read(buf)
		if n > 0 {
			if _, writeErr := fh.Write(buf[:n]); writeErr != nil {
				return total, writeErr
			}
			hasher.Write(buf[:n])
			total += int64(n)
		}
		if readErr == io.EOF {
			return total, nil
		}
		if readErr != nil {
			return total, readErr
		}
	}
}
